src/model/dla.py

```diff
@@ -0,0 +1,390 @@
+        self.actf = tf.keras.Sequential([
+            BatchNorm2d(momentum=BN_MOMENTUM),
+            layers.ReLU()
+        ])
+        self.conv = DeformConvUnit(channels_out, (3, 3), 1, 'same')
+
+    def call(self, x):
+        x = self.conv(x)
+        x = self.actf(x)
+        return x
+
+
+class IDAUp(tf.keras.Model):
+    def __init__(self, o, channels, up_f, node_type=(DeformConv, DeformConv)):
+        super(IDAUp, self).__init__()
+        for i in range(1, len(channels)):
+            c = channels[i]
+            f = int(up_f[i])
+            proj = node_type[0](o)
+            node = node_type[1](o)
+
+            up = layers.Conv2DTranspose(o, f * 2, strides=f, padding='same', use_bias=False,
+                                        data_format='channels_first', groups=o)
+
+            setattr(self, 'proj_' + str(i), proj)
+            setattr(self, 'up_' + str(i), up)
+            setattr(self, 'node_' + str(i), node)
+
+    def call(self, layers, startp, endp):
+        for i in range(startp + 1, endp):
+            upsample = getattr(self, 'up_' + str(i - startp))
+            project = getattr(self, 'proj_' + str(i - startp))
+            x = project(layers[i])
+            x = upsample(x)
+            layers[i] = x
+            node = getattr(self, 'node_' + str(i - startp))
+            layers[i] = node(layers[i] + layers[i - 1])
+
+
+class DLAUp(tf.keras.Model):
+    def __init__(self, startp, channels, scales, in_channels=None,
+                 node_type=DeformConv):
+        super(DLAUp, self).__init__()
+        self.startp = startp
+        if in_channels is None:
+            in_channels = channels
+        self.channels = channels
+        channels = list(channels)
+        scales = np.array(scales, dtype=int)
+        for i in range(len(channels) - 1):
+            j = -i - 2
+            setattr(self, 'ida_{}'.format(i),
+                    IDAUp(channels[j], in_channels[j:],
+                          scales[j:] // scales[j],
+                          node_type=node_type))
+            scales[j + 1:] = scales[j]
+            in_channels[j + 1:] = [channels[j] for _ in channels[j + 1:]]
+
+    def call(self, layers):
+        out = [layers[-1]]  # start with 32
+        for i in range(len(layers) - self.startp - 1):
+            ida = getattr(self, 'ida_{}'.format(i))
+            ida(layers, len(layers) - i - 2, len(layers))
+            out.insert(0, layers[-1])
+        return out
+
+
+DLA_NODE = {
+    'dcn': (DeformConv, DeformConv)
+}
+
+
+class DLASeg(base_model.BaseModel):
+    def __init__(self, num_layers, heads, head_convs, opt):
+        super().__init__(heads, head_convs, 1, opt=opt)
+        down_ratio = 4
+        self.opt = opt
+        self.node_type = DLA_NODE[opt.dla_node]
+        print('Using node type:', self.node_type)
+        self.first_level = int(np.log2(down_ratio))
+        self.last_level = 5
+        self.base = globals()['dla{}'.format(num_layers)](opt=opt)
```

Comment:
It might be a matter of personal taste, but in general I prefer to keep introspection to a minimum. Consider using a custom dictionary for this instead.